x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/IndexLifecycleService.java

```diff
@@ -39,9 +41,11 @@
  * A service which runs the {@link LifecyclePolicy}s associated with indexes.
  */
 public class IndexLifecycleService extends AbstractComponent
-        implements ClusterStateListener, ClusterStateApplier, SchedulerEngine.Listener, Closeable {
+        implements ClusterStateListener, ClusterStateApplier, SchedulerEngine.Listener, Closeable, LocalNodeMasterListener {
```

Comment:
I'm not sure about adding the `LocalNodeMasterListener` could you explain why this is better than doing it in the `ClusterStateListener`?

My worries with `LocalNodeMasterListener`:
* `LocalNodeMasterListener` is called from a `ClusterStateListener` anyway so its only a convenience to listen to when the node is elected and unelected master. I worry that having both a `LocalNodeMasterListener` and a `ClusterStateListener` is going to make the code harder to follow and reason about
* Related to the above point, I worry that splitting this in two could create race conditions in the future if someone adds something to `onMaster()`  which `triggerPolicies()` relies on as I'm not sure that there is a guarantee that `onMaster()` is complete before `clusterChanged()` is called (especially as `onMNAster()~ runs in a different thread. 

I'm fine with this change if it gains us something but I would like to understand what its solving that cannot be solved in `clusterChanged()`?

Note: I think the settings update consumer change is good regardless of whether we use `LocalNodeMasterListener` or not

Reply:
> could you explain why starting and stopping the scheduler in a `LocalNodeMasterListener` is better than doing it in the `ClusterStateListener`?

Doing it with `LocalNodeMasterListener` is being more explicit about the state change. Rather than leaving it up to us to inspect the event state for master changes (transition to or from being master) we let the existing infrastructure deal with it. The nice thing is that `offMaster` is called once and handled for us, rather than having to compare previous state (which, Yannick mentioned to me, we should try and avoid comparing anything with the previous state due to the possibility of missed cluster state updates)

> `LocalNodeMasterListener` is called from a `ClusterStateListener` anyway so its only a convenience to listen to when the node is elected and unelected master. I worry that having both a `LocalNodeMasterListener` and a `ClusterStateListener` is going to make the code harder to follow and reason about

In my opinion it makes it easier, because it's apparent from the method name exactly when the method is called.

Additionally, `LocalNodeMasterListener` forks to the executor so that it doesn't block, which is cleaner execution-wise in my opinion.

> Related to the above point, I worry that splitting this in two could create race conditions in the future if someone adds something to `onMaster()` which `triggerPolicies()` relies on as I'm not sure that there is a guarantee that `onMaster()` is complete before `clusterChanged()` is called (especially as `onMaster()` runs in a different thread).

I think this is better, these should be decoupled, if there is a race condition in the future it's an indication that we need to revisit the way things are triggered. I don't expect us to add much to `onMaster()` though, do you?

> I would like to understand what its solving that cannot be solved in `clusterChanged()`?

Of course, we can do everything in `clusterChanged()`, the benefit of this is that it separate concerns and makes the code intent clearer when anyone unfamiliar with it reads the code. It also makes the `clusterChanged` method smaller and easier to read. Those are the benefits in my mind (and this is also how we handle this same scenario (scheduling and rescheduling a periodic job) in other classes as well (`InternalClusterInfoService`)).

Reply:
Ok then I'm +1 for using `LocalNodeMasterListener`. Thanks for explaining the reasoning.